content/blog/2023/2023-08-31-a-dash-of-dask.md

```diff
@@ -0,0 +1,45 @@
++++
+author = "Noah Novšak"
+date = "2023-08-31"
+draft = false
+title = "A dash of Dask"
+type = "blog"
+thumbImage = ""
+frontPageImage = ""
+blog = ["dask"]
+shortExcerpt = "Updates in the dask project, integrating machine learning methods and enhancing efficiency in a data preprocessor."
+longExcerpt = "Updates in the dask project, integrating machine learning methods and enhancing efficiency in a data preprocessor."
++++
+
+It's been a while since our last progress update on [Dask](/blog/2022/2022-12-13-dask-table), and quite a bit has happened since. Most noticeably, we've got some more visualization widgets working, but I leave that for another blog post. Today, I want to focus on our updated machine-learning. K-Means, PCA, Naive Bayes, Linear, and Logistic Regression are now running. Well, maybe not flawlessly, but hey, they work.
+
+Currently, most of Orange's learners wrap existing `scikit-learn` methods. Extending them to support Dask was as simple as making a helper function that switches between the `sklearn` and `dask_ml` implementations depending on the data received (with a couple of caveats).
+
+Ideally, this kind of data handling should be separate from the canvas and widgets. Both to facilitate development and to not obtrude the end user. Also, we need the learners to maintain backward compatibility because Orange is quite a large project. These requirements mean some _magic_ has to be done behind the scenes to account for the subtle differences between scikit-learn and dask-ml. So, to anyone trying to use one of the learners mentioned above in unpredicted ways, you have been warned.
+
+Additionally, some decisions had to be made about when operations should be computed and what should be stored in memory. Consider, for example, this scenario:
+
+You have made some changes to logistic regression, so you run some tests to ensure everything works, but it takes a while to finish. Waiting is not surprising since you are processing a large amount of data, which usually takes time. But sometimes you wonder if you might be waiting a little too long. So you think, 'Maybe I've made a mistake'. You decide to go through every phase of the computation painstakingly and ensure nothing takes an unreasonable amount of time. After a while, you happen upon a preprocessor that, at first glance, seems to be doing little computing but is taking inordinately long to finish! The preprocessor is called RemoveNaNColumns, and its job is to eliminate columns with too many missing values like so:
```

Comment:
Here you switch to the second person, "you".

Make sure that the user understand that this precise scenario happened to us and the it was bugging us for quite a while (here you can be pretty specific).